Replace occurences of alias_method_chain with their Module#prepend counterpart

kirs opened the pull request:
Earlier @fbernier proposed a cool idea of replacing `alias_method_chain` with `Module#prepend` in #17394.
I guess it's now time to merge it :rocket:
/cc @rafaelfranca

fbernier commented:
Thanks for this, looks like some tests are failing. I will be putting some more work on it.

kaspth commented on activesupport/lib/active_support/core_ext/marshal.rb:
```diff
@@ -1,9 +1,6 @@
-    def load_with_autoloading(source)
-      load_without_autoloading(source)
+module MarshalWithAutoloading
+    def load(source)
```
Outdent this by two spaces

kaspth commented on activesupport/lib/active_support/core_ext/marshal.rb:
```diff
@@ -15,7 +12,10 @@ def load_with_autoloading(source)
+end
 
-    alias_method_chain :load, :autoloading
+Marshal.module_eval do
```
Would using class eval remove the need to use "<< self"

sgrif replied:
`prepend` is public in 2.2, you can just call `Marshal.prepend` directly.

sgrif replied:
Ah nevermind I see what you're doing

sgrif replied:
How about `Marshal.singleton_class.prepend(MarshalWithAutoloading)`?

kaspth replied:
:+1: for Sean's suggestion.

kaspth commented on activesupport/lib/active_support/core_ext/range/include_range.rb:
```diff
@@ -9,15 +7,15 @@ class Range
+      super(value.first) && value.last.send(operator, last)
     else
-      include_without_range?(value)
+      super(value)
```
No need to pass the arg, `super` will do.

kaspth commented on activesupport/test/core_ext/marshal_test.rb:
```diff
@@ -15,7 +15,7 @@ def teardown
     sanity_data.each do |obj|
       dumped = Marshal.dump(obj)
-      assert_equal Marshal.load_without_autoloading(dumped), Marshal.load(dumped)
+      assert_nothing_raised { Marshal.load(dumped) }
```
I'm not sure this is the right change, but I got no suggestions right now.

yuki24 replied:
Do this:

```ruby
assert_equal Marshal.method(:load).super_method.call(dumped), Marshal.load(dumped)
```

kaspth commented:
Should we follow through with a deprecation on `alias_method_chain`?

kirs commented:
Thanks for review! I've pushed some updated parts.

@kaspth it would be great to deprecate `alias_method_chain`, but we need first to update the JSON encoding too: https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/json.rb#L41

I tried it and I also tried to fix the failing test, but I had no success yet. I will spend some more time on it tomorrow :yellow_heart:

kaspth commented:
Well, the question was more aimed at any Core Team members that might have happened to listen in. But yes, do look more into it :smile:

--
Kasper

> Den 19/03/2015 kl. 23.23 skrev Kir Shatrov <notifications@github.com>:
> 
> Thanks for review! I've pushed some updated parts.
> 
> @kaspth it would be great to deprecate alias_method_chain, but we need first to update the JSON encoding too: https://github.com/rails/rails/blob/master/activesupport/lib/active_support/core_ext/object/json.rb#L41
> 
> I tried it and I also tried to fix the failing test, but I had no success yet. I will spend some more time on it tomorrow 
> 
> —
> Reply to this email directly or view it on GitHub.
>

fbernier commented:
The tests are failing [here](https://github.com/rails/rails/blob/33302fdd14d6e82293d1d8ea151af1197e7d74e2/activesupport/test/core_ext/range_ext_test.rb#L109-L114) because of a SystemStackError. After looking into it, I highly suspect it being a ruby bug and I've found [this issue](https://bugs.ruby-lang.org/issues/10847) which sounds like [our problem](https://github.com/ruby/ruby/blob/ca6b174078fa15f33655be704d9409fdbc4f9929/range.c#L1195).

If that's the case, this particular modification will have to wait again I guess.

yuki24 commented:
@fbernier you are right. We can't just replace `#alias_chain_method` with `#prepend` at this point due to that Ruby bug and this only works on Ruby 2.3-head since it's not backported the 2.2-head.

Regarding `#to_json`, you can do something like this:

```ruby
module ActiveSupport
  module CoreExt
    module ToJsonWithActiveSupportEncoder
      def to_json(options = nil)
        if options.is_a?(::JSON::State)
          # Called from JSON.{generate,dump}, forward it to JSON gem's to_json
          super
        else
          # to_json is being invoked directly, use ActiveSupport's encoder
          ActiveSupport::JSON.encode(self, options)
        end
       end
     end
   end
 end
 
[Object, Array, FalseClass, Float, Hash, Integer, NilClass, String, TrueClass, Enumerable].reverse_each do |klass|
  klass.prepend ActiveSupport::CoreExt::ToJsonWithActiveSupportEncoder
end
```

The key here is to use **`#reverse_each`**, not `#reach`.

kirs committed:
Use Module#prepend instead of alias_method_chain
Thanks @fbernier for suggestion! <3
At this moment we can use Module#prepend in all all cases
except of Range because of the bug [1] in MRI 2.2
[1] https://bugs.ruby-lang.org/issues/10847

kirs commented:
Thanks @yuki24!

I updated all the code.
I suggest that now we deprecate `alias_method_chain`, migrate to `Module#prepend` everywhere except of `Range` and then patch `Range` as soon as the bug in MRI will be fixed.

@kaspth @rafaelfranca @fbernier what do you think?

yuki24 commented on activesupport/lib/active_support/core_ext/marshal.rb:
```diff
@@ -1,21 +1,17 @@
-      else
-        raise exc
-      end
+module MarshalWithAutoloading
```
I think this module should have the `ActiveSupport` namespace.

rafaelfranca replied:
:+1:

rafaelfranca commented:
> I suggest that now we deprecate alias_method_chain, migrate to Module#prepend everywhere except of Range and then patch Range as soon as the bug in MRI will be fixed.

:+1:

rafaelfranca commented:
But lets add the deprecation in a different PR.

kirs commented:
@rafaelfranca updated with the namespaced module. I hope it's ready for merge :rocket:

rafaelfranca commented on activesupport/lib/active_support/core_ext/marshal.rb:
```diff
@@ -1,9 +1,7 @@
-    def load_with_autoloading(source)
-      load_without_autoloading(source)
+module ActiveSupport
+  module MarshalWithAutoloading
```
`# :nodoc:`

rafaelfranca commented on activesupport/lib/active_support/core_ext/object/json.rb:
```diff
@@ -26,22 +25,27 @@
+
+module ActiveSupport
+  module CoreExt
+    module ToJsonWithActiveSupportEncoder
```
`# :nodoc:`

yuki24 replied:
@kirs @rafaelfranca I should've mentioned this too. Which namespace is more ideal, `ActiveSupport::CoreExt::...` or just `ActiveSupport::...`?

rafaelfranca replied:
Just changed to `ActiveSupport::`. 7c7d9dcf5a966a5490fe99230ab9513d73df9630

yuki24 replied:
:+1:

rafaelfranca commented:
I commented but don't need to change. I'll do.
